addons/mail/models/res_partner.py

```diff
@@ -116,8 +116,9 @@ def im_search(self, name, limit=20):
                 WHERE P.name ILIKE %s
                     AND P.id NOT IN %s
                     AND U.active = 't'
+                    AND %s NOT IN (SELECT gid FROM res_groups_users_rel WHERE uid = U.id)
                 LIMIT %s
-            """, ("%s seconds" % DISCONNECTION_TIMER, "%s seconds" % AWAY_TIMER, name, tuple(excluded_partner_ids), limit))
+            """, ("%s seconds" % DISCONNECTION_TIMER, "%s seconds" % AWAY_TIMER, name, tuple(excluded_partner_ids), self.env.ref('base.group_portal').id, limit))
```

Comment:
Instead of "not in portal", it should be "in internal users" specifically (`base.group_user`).

This should be doable by simply checking `share = False` in SQL as it is a stored field.